### PR TITLE
1.14.5 fees

### PR DESCRIPTION
**Pull request by rnicoll**
Opening this early so if people are looking at fees, we have a single branch to stage on.

### CODE REVIEW

**Comment by ReverseControl** [comment]
@rnicoll The reason ``fundrawtransaction.py `` fails is because ``round_tx_size()`` is not defined. In particular:

https://github.com/dogecoin/dogecoin/blob/4c93783ab64ba96e60b7ea37202c072c66f3f544/qa/rpc-tests/test_framework/util.py#L519-L521

does not exist in 1.14.5. I think this is because we no longer round up in kB to calculate fees. I would think removing it and counting actual number of bytes would address that test, but not sure since I am not familiar with the rpc-tests.

File: qa/rpc-tests/fundrawtransaction.py [hunk 1]
```diff
@@ -639,10 +639,9 @@ def run_test(self):
         outputs = {self.nodes[3].getnewaddress() : 1}
```
Comment (chalermchai29) [inline qa/rpc-tests/fundrawtransaction.py:639]: rawtx = self.nodes[3].createrawtransaction(inputs, outputs)
        result = self.nodes[3].fundrawtransaction(rawtx) # uses min_relay_tx_fee (set by settxfee)
        # Note TX fees in 1.14.5 are 10x the required minimum
        result2 = self.nodes[3].fundrawtransaction(rawtx, {"feeRate": 2*10*min_relay_tx_fee})
        result3 = self.nodes[3].fundrawtransaction(rawtx, {"feeRate": 10*10*min_relay_tx_fee})
        result_fee_rate = result['fee'] * 1000 / round_tx_size(count_bytes(result['hex']))
        assert_fee_amount(result2['fee'], count_bytes(result2['hex']), 2 * result_fee_rate)
        assert_fee_amount(result3['fee'], count_bytes(result3['hex']), 10 * result_fee_rate)

**Review by chalermchai29 — CHANGES_REQUESTED** [review]

File: qa/rpc-tests/test_framework/util.py [hunk 2]
```diff
@@ -509,15 +509,12 @@ def random_transaction(nodes, amount, min_fee, fee_increment, fee_variants):
 
 def assert_fee_amount(fee, tx_size, fee_per_kB):
     """Assert the fee was in range"""
-    target_fee = fee_per_kB / 1000
```
Comment (chalermchai29) [inline qa/rpc-tests/test_framework/util.py:512]: target_fee = tx_size * fee_per_kB / 1000

File: qa/rpc-tests/test_framework/util.py [hunk 3]
```diff
@@ -509,15 +509,12 @@ def random_transaction(nodes, amount, min_fee, fee_increment, fee_variants):
...
+    target_fee = tx_size * fee_per_kB / 1000
     if fee < target_fee:
```
Comment (chalermchai29) [inline qa/rpc-tests/test_framework/util.py:513]: if fee < target_fee:
        raise AssertionError("Fee of %s DOGE too low! (Should be %s DOGE)"%(str(fee), str(target_fee)))
    # allow the wallet's estimation to be at most 2 bytes off
    if fee > (tx_size + 2) * fee_per_kB / 1000:
        raise AssertionError("Fee of %s DOGE too high! (Should be %s DOGE)"%(str(fee), str(target_fee)))

File: qa/rpc-tests/test_framework/util.py [hunk 4]
```diff
@@ -509,15 +509,12 @@ def random_transaction(nodes, amount, min_fee, fee_increment, fee_variants):
...
-        raise AssertionError("Fee of %s BTC too low! (Should be %s BTC)"%(str(fee), str(target_fee)))
```
Comment (chalermchai29) [inline qa/rpc-tests/test_framework/util.py:514]: raise AssertionError("Fee of %s DOGE too low! (Should be %s DOGE)"%(str(fee), str(target_fee)))
    # allow the wallet's estimation to be at most 2 bytes off
    if fee > (tx_size + 2) * fee_per_kB / 1000:
        raise AssertionError("Fee of %s DOGE too high! (Should be %s DOGE)"%(str(fee), str(target_fee)))

File: src/amount.cpp [hunk 5]
```diff
@@ -25,21 +25,6 @@ CAmount CFeeRate::GetFee(size_t nBytes_) const
     assert(nBytes_ <= uint64_t(std::numeric_limits<int64_t>::max()));
     int64_t nSize = int64_t(nBytes_);
 
-    // Dogecoin: Round up to the nearest 1000 bytes so we get round tx fees
```
Comment (chalermchai29) [inline src/amount.cpp:28]: assert(nBytes_ <= uint64_t(std::numeric_limits<int64_t>::max()));
    int64_t nSize = int64_t(nBytes_);

Comment (chalermchai29) [reply src/amount.cpp:28]: assert(nBytes_ <= uint64_t(std::numeric_limits<int64_t>::max()));
    int64_t nSize = int64_t(nBytes_);

**Review by chalermchai29 — APPROVED** [review]

File: qa/rpc-tests/fundrawtransaction.py (hunk 1 above)
Comment (ReverseControl) [reply qa/rpc-tests/fundrawtransaction.py:639]: @charlescharles Two observations:   

1. What branch did you develop on? @rnicoll  branch does not have ``round_tx_size`` in the latest commit on this work?
2.   It appears you have developed on top of 531f103ea5fe96798eda49b642f136f8b103cc10. Could you please update from the latest commit on this work?

https://github.com/dogecoin/dogecoin/blob/81d5fa40e8d4820bc1363f646bfa27294014a5cd/qa/rpc-tests/abandonconflict.py#L19

``"-minrelaytxfee=0.00001"`` has two more zeroes than needed.

**Comment by rnicoll** [comment]
@ReverseControl Thanks, fixed up fundrawtransaction.py now

**Comment by rnicoll** [comment]
@chalermchai29 I'm confused by a lot of your comments, but I think they're trying to revert back to the old version before we reduced fees?

**Comment by ReverseControl** [comment]
@rnicoll See my comment on first suggestion. Yes, you are correct. He is working from an old branch.

**Comment by ReverseControl** [comment]
@rnicoll If you could explain to me, offline, how to compute all the fees and how the test framework works I can take care of all these erros.

**Comment by rnicoll** [comment]
> @rnicoll If you could explain to me, offline, how to compute all the fees and how the test framework works I can take care of all these erros.

I think I've fixed the errors, next problem is the values in https://github.com/dogecoin/dogecoin/pull/2485/commits/af5e7b6ce5f693a548f82ba59b1530c045da1bd8 are very slightly lower than I'd expect. I'd expected:

* 999.9775
* 999.775
* 999.9775

It's likely the transactions are a byte bigger than I've expected, but not sure. If you have the energy to calculate the amounts taking into account transaction actual size, it would be appreciated.

**Comment by ReverseControl** [comment]
@rnicoll Even if for my own reference, I would very much like to understand all the fees. Who pays them, who receives them, when, how, and why. A chart and an explanation would allow me to fly through fixing RPC errors, now and in the future. Or just a chat about them.

**Comment by rnicoll** [comment]
> @rnicoll Even if for my own reference, I would very much like to understand all the fees. Who pays them, who receives them, when, how, and why. A chart and an explanation would allow me to fly through fixing RPC errors, now and in the future. Or just a chat about them.

@ReverseControl You've seen https://github.com/dogecoin/dogecoin/discussions/2347 right?

Basics, in the fees from 1.14.5:

* Basic transaction fee is 0.01 DOGE * size of the transaction in kilobytes. For example for a typical transaction of 225 bytes, it's 0.225 kB, so the fee is 0.01 * 0.225 = 0.00225 DOGE.
* For transactions with an output below the dust threshold (0.01 from 1.14.5 I believe - check the discussion link above for the correct value, and if you can verify the tests cover it properly that'd be great) then a 0.01 DOGE penalty fee is added.
* Fees are typically paid by the sender. For example if I send 10 DOGE with a 0.225 kB transaction, I actually input 10.00225 DOGE.
* Transaction inputs are outputs from previous transactions, and are consumed entirely when this are used. Anything left over is then sent to a new address as change. The fee is the difference between the inputs and the outputs.
  * For example, if the inputs are 11 DOGE, the fee 0.00225 DOGE and the transaction sends 10 DOGE, then 0.99775 (11 DOGE - 10 DOGE - 0.00225 DOGE) sent to a change output. That means the transaction has two outputs, one of 10 DOGE, the other of 0.99775 DOGE.

There's then weird corner cases like relay required fees and block inclusion and free transaction space, but the above covers 99% of cases.

**Comment by ReverseControl** [comment]
@rnicoll Yes, I have read that. 

So, questions:

1.  Who pays the relay fees? To whom? And, when?
2. What is the dust limit? I have seen it quite a bit in the code.
3. For the last example, who owns the new created wallet to store the change?


The corner cases are the most interesting cases, because those are the ones we have to absolutely get right.

**Comment by rnicoll** [comment]
@ReverseControl Relay fees go to the miners, they're just minimums that nodes use to decide if to relay. The idea is that nodes don't bother to relay anything that wouldn't be mined (because it's just network spam), rather than they're paid to the relaying nodes.

Dust limit is the amount below which an output is considered ridiculously small. Typically any change below dust is just given to the miner as additional fee, because the output isn't worth the data to store it.

The sender owns the change address. There's a much longer write up on change addresses at https://support.blockchain.com/hc/en-us/articles/360018566291-Change-addresses

**Comment by ReverseControl** [comment]
@rnicoll  Nodes do not have coins, who pays the relaying fee to the miners? 

Good to know what dust limit is and how it is used. I took a look at the link. That makes more sense.

**Comment by rnicoll** [comment]
> @rnicoll Nodes do not have coins, who pays the relaying fee to the miners?
> 
> Good to know what dust limit is and how it is used. I took a look at the link. That makes more sense.

@ReverseControl Relay fees is just a threshold value, it's not its own fee type. So they're a value which the fee paid by the sender to the miner must exceed, for nodes to be prepared to relay the transaction. That make any sense?

**Comment by rnicoll** [comment]
> 
> 
> @rnicoll I think I understand, so in your first bullet example: the fee paid by the sender is `0.01 * 0.225 KB = 0.00225 DOGE` because this value is greater than the Relayfee of `.001 DOGE`nodes are inclined to relay it to miners, but if a block had size 90 bytes (let's pretend that is possible) `0.01 * 0.09 KB = 0.0009 DOGE` then nodes would be inclined to decline sending that transaction to miners?

Very close - relay fee is also in DOGE per kilobyte, so for a 0.225kB tx and a relay fee of 0.001 DOGE per kilobyte, the fee for the transaction must be at least 0.000225 (yes 1/10th of the fee used) for it to be relayed.

So in your example if it was _below_ 0.0009 DOGE it wouldn't be relayed.

**Comment by ReverseControl** [comment]
@rnicoll You mean ``.00009``, right? ``.001*.09 = .00009``, not ``0.0009``.

So, in essence so long as ``Basic Transaction > Relay Fee``, noted these are both fees per KB,  nodes will be inclined to pass your tx to miners?

**Comment by rnicoll** [comment]
> 
> 
> @rnicoll You mean `.00009`, right? `.001*.09 = .00009`, not `0.0009`.
> 
> So, in essence so long as `Basic Transaction > Relay Fee`, noted these are both fees per KB, nodes will be inclined to pass your tx to miners?

Yes to 0.0009, sorry. `Basic Transaction >= Relay Fee` technically but I think you got the idea

**Comment by ReverseControl** [comment]
@rnicoll  Great. For now that is enough. I will ask later about free space and block inclusion fees.

I will take a look https://github.com/dogecoin/dogecoin/commit/af5e7b6ce5f693a548f82ba59b1530c045da1bd8 now.

**Comment by ReverseControl** [comment]
@rnicoll @patricklodder Where are all the methods of ``self.nodes[0]`` documented? or whatever the index happens to be. I can't find them. I need the list.

**Comment by ReverseControl** [comment]
@rnicoll @patricklodder  what is WITNESS_SCALE_FACTOR? What is it used for? Also, what is DEFAULT_BYTES_PER_SIGOP? What is it used for?

**Comment by patricklodder** [comment]
`WITNESS_SCALE_FACTOR` is inherited from Bitcoin and a control to increase effective blocksize limits without changing actual blocksize limits by introducing Segregated Witness functionality, which is not enabled on Dogecoin. In [earlier discussions regarding the soft fork plan](https://github.com/dogecoin/dogecoin/discussions/2264) it was brought up that this is overly complicated and while it served Bitcoin with an under-the-hood blocksize increase, for Dogecoin this does not make much sense because we have no small/big block debate in the sense of how contentious that was (and still is, kind of) for Bitcoin.

`DEFAULT_BYTES_PER_SIGOP` is an anti-spam measure that prevents the situation where an attacker could craft small transactions with tons of operations on the verification script, that would cost enormous time to process and basically DoS block verification throughout the network. It limits how many operations are allowed per transaction byte. This was a 2010 CVE: CVE-2010-5138

**Comment by patricklodder** [comment]
> Where are all the methods of `self.nodes[0]` documented?

Those are the RPC methods, you can see them with `dogecoin-cli help`

**Comment by ReverseControl** [comment]
@rnicoll Though a priority it might not be, given the monetary cost is so minimal, I am rather pedantic about details like this; that is, doing things correctly in the strictest logical sense of the word correct. Would it not be appropriate to fix this issue by first signing the transaction for real and then calculating the fee with the actual transaction that will be recorded in the blockchain? As opposed to making assumptions, calculating the fee, and then signing the transaction that will become the legitimate transaction after the fact.

**Comment by rnicoll** [comment]
> 
> 
> @rnicoll Though a priority it might not be, given the monetary cost is so minimal, I am rather pedantic about details like this; that is, doing things correctly in the strictest logical sense of the word correct. Would it not be appropriate to fix this issue by first signing the transaction for real and then calculating the fee with the actual transaction that will be recorded in the blockchain? As opposed to making assumptions, calculating the fee, and then signing the transaction that will become the legitimate transaction after the fact.

That's the issue - we can't know the fee until after signing, and we can't update the fee after signing without breaking the signature. Instead we use a fee value which is long enough to cover the biggest signature we can expect.

**Review by edtubbs — APPROVED** [review]

File: qa/rpc-tests/fundrawtransaction.py [hunk 6]
```diff
@@ -198,11 +197,10 @@ def run_test(self):
         ############################################################
         # test a fundrawtransaction with a provided change address #
         ############################################################
-        utx = get_unspent(self.nodes[2].listunspent(), 50)
+        utx = get_unspent(self.nodes[2].listunspent(), 5)
 
         inputs  = [ {'txid' : utx['txid'], 'vout' : utx['vout']} ]
-        # Dogecoin: Reduce this output so the fee doesn't leave us with no change
-        outputs = { self.nodes[0].getnewaddress() : Decimal(25) }
+        outputs = { self.nodes[0].getnewaddress() : Decimal(4.0) }
```
Comment (michilumin) [inline qa/rpc-tests/fundrawtransaction.py:203]:
```suggestion
        outputs = { self.nodes[0].getnewaddress() : Decimal(2.5) }
```

**Review by michilumin — APPROVED** [review]
oof. Pretty sure this is good to go. Lots of math to follow but - seems sane.

**Pull request merged**